### ISSUE TITLE
QueryTrait::indexBy works unoptimal in some cases

### ISSUE THREAD

**Issue by creocoder** [yiisoft/yii2#9342]
Consider case:

``` php
static::find()
    ->with('group')
    ->indexBy(function(Plan $row) {
        return $row->group->slug . '-' . $row->interval;
    })
    ->all();
```

This will lead to 2 + N queries instead of 2 queries:

```
SELECT * FROM `plan`;
SELECT * FROM `plan_group` WHERE `id`=3; // excess query
SELECT * FROM `plan_group` WHERE `id`=2; // excess query
SELECT * FROM `plan_group` WHERE `id`=1; // excess query
SELECT * FROM `plan_group` WHERE `id` IN (1, 2, 3);
```

`QueryTrait::indexBy` should work after eager loading happen. So queries should be:

```
SELECT * FROM `plan`;
SELECT * FROM `plan_group` WHERE `id` IN (1, 2, 3);
```

Even if replace `with()` to `joinWith()` issue still happen.

**Comment by cebe** [comment]
here is why:

Models are populated including the indexBy application:
https://github.com/yiisoft/yii2/blob/master/framework/db/ActiveQuery.php#L220
and only afterwards, with is considered:
https://github.com/yiisoft/yii2/blob/master/framework/db/ActiveQuery.php#L225

**Comment by creocoder** [comment]
@cebe Ok, can we fix it? Because such behavior makes `indexBy` less usable. Also its unexpected behavior. I understand that it caused by our architecture.

**Comment by cebe** [comment]
not sure, need to look into that.

**Comment by SamMousa** [comment]
I've implemented this, but noticed something when I created a test for it.
This is (still) not supported:
````
Order::find()->indexBy('customer.id')->with('customer')->all();
````

We could choose to use `ArrayHelper::getValue` for resolving the value, this might influence performance.

@cebe What's your opinion on that?

**Comment by cebe** [comment]
@SamMousa the order of creating models and populating `with` is still against this use case:
https://github.com/yiisoft/yii2/blob/726d4d1b6b1fe6eed6b272d665a2a217670acdfd/framework/db/ActiveQuery.php#L210-L216

Not sure if we can easily change that.

**Comment by SamMousa** [comment]
We can: https://github.com/yiisoft/yii2/issues/13891

**Comment by cebe** [comment]
Oh nice, have not seen that yet.

**Comment by SamMousa** [comment]
Okay, when you had a chance to look at it; plz reconsider my question ;-)

**Comment by cebe** [comment]
> We could choose to use ArrayHelper::getValue for resolving the value, this might influence performance.

I think it is fine to do so.

**Comment by SamMousa** [comment]
Let's see if tests still pass.

**Comment by SilverFire** [comment]
Fixed in 807945629
